ambry-frontend/src/main/java/com/github/ambry/frontend/GetBlobHandler.java

```diff
@@ -0,0 +1,253 @@
+    if (subResource == null || subResource.equals(SubResource.Segment)) {
+      group = frontendMetrics.getBlobMetricsGroup;
+    } else {
+      switch (subResource) {
+        case BlobInfo:
+          group = frontendMetrics.getBlobInfoMetricsGroup;
+          break;
+        case UserMetadata:
+          group = frontendMetrics.getUserMetadataMetricsGroup;
+          break;
+        case Replicas:
+          group = frontendMetrics.getReplicasMetricsGroup;
+          break;
+      }
+    }
+    return group;
+  }
+
+  private class CallbackChain {
+    private final RestRequest restRequest;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<ReadableStreamChannel> finalCallback;
+    private final RestRequestMetricsGroup metricsGroup;
+    private final RequestPath requestPath;
+    private final SubResource subResource;
+    private final GetBlobOptions options;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param restResponseChannel the {@link RestResponseChannel}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     */
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        RestRequestMetricsGroup metricsGroup, RequestPath requestPath, SubResource subResource, GetBlobOptions options,
+        Callback<ReadableStreamChannel> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.metricsGroup = metricsGroup;
+      this.requestPath = requestPath;
+      this.subResource = subResource;
+      this.options = options;
+      this.finalCallback = finalCallback;
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      securityService.processRequest(restRequest, securityProcessRequestCallback());
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link IdConverter#convert} to convert the incoming
+     * ID if required.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback() {
+      return buildCallback(metrics.getBlobSecurityProcessRequestMetrics, result -> {
+        String blobIdStr = requestPath.getOperationOrBlobId(false);
+        idConverter.convert(restRequest, blobIdStr, idConverterCallback());
+      }, restRequest.getUri(), LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
+     */
+    private Callback<String> idConverterCallback() {
+      return buildCallback(metrics.getBlobIdConversionMetrics, convertedBlobId -> {
```

Comment:
Can we avoid lambdas and put this code in a named function ?

Reply:
that would be the next PR that would move handler logic away from callbacks and into the completablefuture world.